### PR TITLE
Optimization

### PR DESCRIPTION
**Pull request by martinateruzzi**
Optimization algorithm for best switches configuration:
- switches state as `True`/`False` 
- switches are considered `True` if they allow the flow within the graph, `False` otherwise
- fitness computed taking into account number of actions (wrt initial condition), total final service, and total number of surviving nodes after the perturbation

### CODE REVIEW

File: grape/fault_diagnosis.py [hunk 1]
```diff
@@ -120,6 +119,143 @@ def check_before(self):
...
+        T = GeneralGraph()
+        T.add_nodes_from(self.G) 
+        for (u, v, d) in self.G.edges(data=True):
+            T.add_edge(u, v, father_condition=d['father_condition'],
+                weight=d['weight'])
+        nx.set_node_attributes(T, self.G.initial_service,
+            name='initial_service')
+        nx.set_node_attributes(T, self.G.perturbation_resistant,
+            name='perturbation_resistant')
+        nx.set_node_attributes(T, self.G.type, name='type')
```
Comment (ndem0) [inline grape/fault_diagnosis.py:150]: Is this just a copy of the original graph? In the case, `T = self.copy()` should do the job

File: grape/fault_diagnosis.py [hunk 2]
```diff
@@ -120,6 +119,143 @@ def check_before(self):
...
+
+            # Apply crossover and mutation on the offspring
+            for child1, child2 in zip(offspring[::2], offspring[1::2]):
+                if random.random() < 0.5:
+                    toolbox.mate(child1, child2)
+                    del child1.fitness.values
+                    del child2.fitness.values
+    
+            for mutant in offspring:
+                if random.random() < 0.5:
```
Comment (ndem0) [inline grape/fault_diagnosis.py:238]: I would move the hardcoded threshold to the beginning of the method (maybe in the future also to `param` to give the user the possibility to change such value)

File: grape/fault_diagnosis.py [hunk 3]
```diff
@@ -120,6 +119,143 @@ def check_before(self):
...
+            g = g + 1
+
+            # Select the next generation individuals
+            offspring = toolbox.select(pop, 5)
+            # Clone the selected individuals
+            offspring = list(map(toolbox.clone, offspring))
+
+            # Apply crossover and mutation on the offspring
+            for child1, child2 in zip(offspring[::2], offspring[1::2]):
+                if random.random() < 0.5:
```
Comment (ndem0) [inline grape/fault_diagnosis.py:232]: I would move the hardcoded threshold to the beginning of the method (maybe in the future also to `param` to give the user the possibility to change such value)

File: grape/fault_diagnosis.py [hunk 4]
```diff
@@ -120,6 +119,143 @@ def check_before(self):
...
+        fitnesses = fits
+        result = []
+        # Begin the evolution
+
+        while g < params['ngen']:
+            # A new generation
+            g = g + 1
+
+            # Select the next generation individuals
+            offspring = toolbox.select(pop, 5)
```
Comment (ndem0) [inline grape/fault_diagnosis.py:226]: I would move the hardcoded threshold to the beginning of the method (maybe in the future also to `param` to give the user the possibility to change such value)

File: grape/fault_diagnosis.py (hunk 1 above)
Comment (martinateruzzi) [reply grape/fault_diagnosis.py:150]: It is a copy of the original graph, in the sense same number of nodes and same edges, but with very few attributes, namely `initial_service`, `perturbation_resistant` and `type`.

**Pull request merged**